OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbySavedSettingsBrowserLogic.cs

```diff
@@ -0,0 +1,351 @@
+					if (CanSave())
+						Save();
+					return true;
+				};
+				saveList.Bounds.Height -= saveWidgets.Bounds.Height;
+				saveWidgets.IsVisible = () => true;
+
+				var saveButton = panel.Get<ButtonWidget>("SAVE_BUTTON");
+				saveButton.OnClick = Save;
+				saveButton.IsVisible = () => true;
+				saveButton.IsDisabled = () => !CanSave();
+			}
+			else
+			{
+				panel.Get("LOAD_TITLE").IsVisible = () => true;
+
+				var loadButton = panel.Get<ButtonWidget>("LOAD_BUTTON");
+				loadButton.OnClick = Load;
+				loadButton.IsVisible = () => true;
+				loadButton.IsDisabled = () => selectedSave == null;
+			}
+
+			var renameButton = panel.Get<ButtonWidget>("RENAME_BUTTON");
+			renameButton.IsDisabled = () => selectedSave == null;
+			renameButton.OnClick = () =>
+			{
+				var initialName = Path.GetFileNameWithoutExtension(selectedSave);
+				var invalidChars = Path.GetInvalidFileNameChars();
+
+				ConfirmationDialogs.TextInputPrompt(
+					"Rename Save",
+					"Enter a new file name:",
+					initialName,
+					onAccept: newName => Rename(initialName, newName),
+					onCancel: null,
+					acceptText: "Rename",
+					cancelText: null,
+					inputValidator: newName =>
+					{
+						if (newName == initialName)
+							return false;
+
+						if (string.IsNullOrWhiteSpace(newName))
+							return false;
+
+						if (newName.IndexOfAny(invalidChars) >= 0)
+							return false;
+
+						if (File.Exists(SettingsFilePath(newName)))
+							return false;
+
+						return true;
+					});
+			};
+
+			var deleteButton = panel.Get<ButtonWidget>("DELETE_BUTTON");
+			deleteButton.IsDisabled = () => selectedSave == null;
+			deleteButton.OnClick = () =>
+			{
+				ConfirmationDialogs.ButtonPrompt(
+					title: "Delete selected lobby settings save?",
+					text: "Delete '{0}'?".F(Path.GetFileNameWithoutExtension(selectedSave)),
+					onConfirm: () =>
+					{
+						Delete(selectedSave);
+
+						if (!saves.Any() && !isSavePanel)
+						{
+							Ui.CloseWindow();
+						}
+						else
+							SelectFirstVisible();
+					},
+					confirmText: "Delete",
+					onCancel: () => { });
+			};
+
+			LoadExistingSaves(newTemplate, savedTemplate);
+			SelectFirstVisible();
+		}
+
+		void SelectFirstVisible()
+		{
+			Select(isSavePanel ? null : saves.FirstOrDefault());
+		}
+
+		void Select(string savePath)
+		{
+			selectedSave = savePath;
+			if (isSavePanel)
+				saveTextField.Text = savePath == null ? defaultSaveFilename :
+					Path.GetFileNameWithoutExtension(savePath);
+		}
+
+		void LoadExistingSaves(ScrollItemWidget newTemplate, ScrollItemWidget savedTemplate)
+		{
+			saveList.RemoveChildren();
+
+			if (isSavePanel)
+			{
+				var item = ScrollItemWidget.Setup(newTemplate,
+					() => selectedSave == null || !saves.Contains(SettingsFilePath(saveTextField.Text)),
+					() => { });
+				saveList.AddChild(item);
+			}
+
+			if (Directory.Exists(baseSavePath))
+			{
+				var saveFiles = Directory.GetFiles(baseSavePath, "*.yaml", SearchOption.AllDirectories)
```

Comment:
I think we should stick with `TopDirectoryOnly` (instead of diving into sub-directories, unless we expect there to be subdirs for some reason that I'm not aware of?), so we aren't reading more than necessary:
```suggestion
				var saveFiles = Directory.GetFiles(baseSavePath, "*.yaml", SearchOption.TopDirectoryOnly)
```

Sorry I missed this initially.
I'm not _trying_ to get you to do even more work, I promise. :P